README.md
# wheelmappeR
This is a wrapper for the Wheelmap API in R


NEWS.md
# wheelmappeR 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(wheelmappeR)

test_check("wheelmappeR")


tests/testthat/test_get_wheelmap_locations.R
context("Get locations")

# TEST INVALID INPUTS
testthat::test_that("test that function returns error if no argument provided", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(),
                         regexp = "You need to specify at least one argument")
})

testthat::test_that("test that function returns error if no API key is provided", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(q="abc"),
                         regexp = "Please provide your API key as an argument.")
})

testthat::test_that("test that function returns error if search term is not a string", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(api_key="123",
                                                             q=1),
                         regexp = "Input for q must be a string.")
})

testthat::test_that("test that function returns error if bbox is not in string format", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(api_key="123",
                                                             bbox=1),
                         regexp = "Input for bbox must be a string with four comma-seperated coordinate values; e.g. \"52.357, 52.677, 13.228, 13.548\"")
})

testthat::test_that("test that function returns error if bbox does not contain four coordinates", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(api_key="123",
                                                             bbox="12.32, 23.32"),
                         regexp = "Input for bbox must be a string with four comma-seperated coordinate values; e.g. \"52.357, 52.677, 13.228, 13.548\"")
})

testthat::test_that("test that function returns error if wheelchair is not a string", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(api_key="123",
                                                             wheelchair=1),
                         regexp = "Input for wheelchair must be one of the following strings \"yes\", \"limited\", \"no\", \"unknown\".")
})

testthat::test_that("test that function returns error if wheelchair is not one of the options", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(api_key="123",
                                                             wheelchair="abc"),
                         regexp = "Input for wheelchair must be one of the following strings \"yes\", \"limited\", \"no\", \"unknown\".")
})

testthat::test_that("test that function returns error if page argument is not numeric", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(api_key="123",
                                                             page="abc"),
                         regexp = "Input for page must be a numeric.")
})

testthat::test_that("test that function returns error if per_page argument is not numeric", {
  testthat::expect_error(wheelmappeR::get_wheelmap_locations(api_key="123",
                                                             per_page="abc"),
                         regexp = "Input for per_page must be a numeric.")
})







R/utils.R
#' @importFrom attempt stop_if_not
#' @importFrom curl has_internet
check_internet <- function(){
  stop_if_not(.x = has_internet(), msg = "Please check your internet connexion")
}

#' @importFrom httr status_code
check_status <- function(res){
  stop_if_not(.x = status_code(res),
              .p = ~ .x == 200,
              msg = "The API returned an error")
}


base_url <- "http://wheelmap.org/api/nodes"

#' Parses content returned by query to the Wheelmap API.
#'
#' \code{parse_wheelmap_content} parses the content sent back by
#' the Wheelmap API to an R list.
#'
#' @param response httr response object
#'
#' @importFrom httr content
#' @importFrom utils data
#' @importFrom jsonlite fromJSON
#'
#' @return R list.
parse_wheelmap_content <- function(response){
  content_text    <- httr::content(response, "text")
  content_parsed  <- jsonlite::fromJSON(content_text)
  return(content_parsed)
}


#' Extracts metadata.
#'
#' \code{extract_wheelmap_metadata} extracts meta data from the response object and the
#' parsed content.
#'
#' @param response httr response object
#' @param content_parsed parsed content of a response to Wheelmap query
#'
#' @return data frame containing meta data related to the query.
extract_wheelmap_metadata <- function(response, content_parsed){
  metadata <- data.frame(total_results = content_parsed$meta$item_count_total,
                         results_returned = content_parsed$meta$item_count,
                         status_code   = response$status_code,
                         request_date  = response$headers$date,
                         request_url   = response$url,
                         stringsAsFactors = FALSE)
  return(metadata)
}


#' Extracts data frame with Wheelmap locations from response object.
#'
#' \code{extract_wheelmap_locations} extracts a data frame containing the Wheelmap sources that
#' matched the request to Wheelmap location endpoint.
#'
#' @param metadata data frame containing meta data related to the request, see extract_wheelmap_metadata.
#' @param content_parsed parsed content of a response to wheelmap query
#'
#' @return data frame containing sources.

extract_wheelmap_locations <- function(metadata, content_parsed){
  empty_df <- data.frame()
  if (metadata$status_code == 200) {
    if(metadata$total_results > 0){
      # from list to data.frame
      df <- do.call(data.frame, content_parsed$nodes)
      #clean df
      df$name <- as.character(df$name)
      df$wheelchair_description <- as.character(df$wheelchair_description)
      df$node_type.identifier <- as.factor(df$node_type.identifier)
      df$id <- as.character(df$id)
      df$category.identifier <- as.factor(df$category.identifier)
      df$street <- as.character(df$street)
      df$housenumber <- as.character(df$housenumber)
      df$city <- as.character(df$city)
      df$postcode <- as.character(df$postcode)
      df$website <- as.character(df$website)
      df$phone <- as.character(df$phone)

      return(df)
    } else {
      warning(paste0("The search was not successful. There were no results",
                     " for your specifications."))
      return(empty_df)
    }
  } else {
    # an error occurred
    warning(paste0("The search resulted in the following error message: ",status_code))
    return(empty_df)
  }
}



R/data_raw/devstuffs.R
install.packages("devtools")
install.packages("roxygen2")
install.packages("usethis")
install.packages("curl")
install.packages("httr")
install.packages("jsonlite")
install.packages("attempt")
install.packages("purrr")
devtools::install_github("r-lib/desc")

library(devtools)
library(usethis)
library(desc)

# Remove default DESC
unlink("DESCRIPTION")
# Create and clean desc
my_desc <- description$new("!new")

# Set your package name
my_desc$set("Package", "wheelmappeR")

#Set your name
my_desc$set("Authors@R", "person('Johannes', 'Mueller', email = 'johannes.m@correlaid.org', role = c('cre', 'aut'))")

# Remove some author fields
my_desc$del("Maintainer")

# Set the version
my_desc$set_version("0.0.0.9000")

# The title of your package
my_desc$set(Title = "Wrapper for the Wheelmap API")
# The description of your package
my_desc$set(Description = "A long description of this super package I'm working on.")
# The urls
my_desc$set("URL", "http://this")
my_desc$set("BugReports", "http://that")
# Save everyting
my_desc$write(file = "DESCRIPTION")

# If you want to use the MIT licence, code of conduct, and lifecycle badge
use_mit_license(name = "Johanne Mueller")
use_code_of_conduct()
use_lifecycle_badge("Experimental")
use_news_md()

# Get the dependencies
use_package("httr")
use_package("jsonlite")
use_package("curl")
use_package("attempt")
use_package("purrr")

# Clean your description
use_tidy_description()

#testing
library(testthat)
usethis::use_testthat()
use_vignette("wheelmappeR")
use_readme_rmd()




R/fun_get_wheelmap_locations_all.R
#' Get list of all locations
#'
#' @param api_key Your API key
#' @param q Filter nodes by a search term
#' @param bbox Filter nodes by a bounding box (bbox=west,south,east,north) as comma separated float numbers wich are longitude, latitude values in degrees.
#' @param wheelchair Filter nodes by a wheelchair status (Valid values: yes, limited, no, unknown)
#'
#' @importFrom attempt stop_if_all
#' @importFrom purrr compact
#' @importFrom jsonlite fromJSON
#' @importFrom httr GET
#' @export
#' @rdname get_wheelmap_locations_all
#'
#' @return the results from the search
#' @examples
#' \dontrun{
#' get_wheelmap_locations_all(api_key = api_key, bbox="52.357, 52.677, 13.228, 13.548")
#' }

get_wheelmap_locations_all <- function(api_key=api_key,q = NULL, bbox = NULL, wheelchair = NULL){

  init <- get_wheelmap_locations(api_key = api_key, q = q, bbox=bbox, wheelchair=wheelchair, page = 1,per_page=500) # max
  total_pages <- floor(init[[1]]$total_results/500)+1

  #Initalize Dataframe
  df <- init[[2]]

  #Loop over pages
  for(i in 2:total_pages){
    new <- get_wheelmap_locations(api_key = api_key, q = q, bbox=bbox, wheelchair=wheelchair, page = i,per_page=500)
    df <- rbind(df, new[[2]], make.row.names=T)
    print(paste0("Processing page ", i," of ",total_pages))
  }

  # Update meta_data
  meta <- init[[1]]
  meta$results_returned <- nrow(df)

  #format output
  output <- list(meta, df)
  names(output) <- c("meta_data", "locations")

  return(output)
}


R/fun_get_wheelmap_locations.R
#' Get list of all locations
#'
#' @param api_key Your API key
#' @param q Filter nodes by a search term
#' @param bbox Filter nodes by a bounding box (bbox=west,south,east,north) as comma separated float numbers wich are longitude, latitude values in degrees.
#' @param wheelchair Filter nodes by a wheelchair status (Valid values: yes, limited, no, unknown)
#' @param page For pagination, what page of the results you are on. Default is 1.
#' @param per_page For pagination, how many results to return per page. Default is 200. Max is 500.
#'
#' @importFrom attempt stop_if_all
#' @importFrom purrr compact
#' @importFrom jsonlite fromJSON
#' @importFrom httr GET
#' @export
#' @rdname get_wheelmap_locations
#'
#' @return the results from the search
#' @examples
#' \dontrun{
#' get_wheelmap_locations(api_key = api_key, bbox="52.35, 52.67, 13.22, 13.54")
#'
#' # Example bbox values for different cities
#' # Berlin "52.3570365, 52.6770365, 13.2288599, 13.5488599"
#' # Dortmund "51.3542273, 51.6742273, 7.3052789, 7.6252789"
#' # Frankfurt "50.0155435, 50.2271408, 8.4718253, 8.8004716"
#' # Hamburg "53.3951118, 53.7394377, 9.7304736, 10.3251401"
#' # Munich "48.0616017, 48.2482197, 11.3607802, 11.7228777"
#' }

get_wheelmap_locations <- function(api_key=NULL,q = NULL, bbox = NULL, wheelchair = NULL, page = NULL, per_page = NULL){

  args <- list(api_key = api_key, q = q, bbox=bbox, wheelchair=wheelchair, page = page,per_page=per_page)

  # Check that at least one argument is not null
  stop_if_all(args[2:6], is.null, "You need to specify at least one argument")

  #check if data type of input arguments is correct
  if (is.null(api_key)) {
    stop("Please provide your API key as an argument.")
  }
  if (!is.null(q) & !is.character(q)) {
    stop("Input for q must be a string.")
  }
  if (!is.null(bbox) & !is.character(bbox)){
    stop("Input for bbox must be a string with four comma-seperated coordinate values; e.g. \"52.357, 52.677, 13.228, 13.548\"")
  }
  if (!is.null(bbox) & !is.character(bbox)){
    stop("Input for bbox must be a string with four comma-seperated coordinate values; e.g. \"52.357, 52.677, 13.228, 13.548\"")
  }

  if(!is.null(bbox)){
    if(is.character(bbox) & length(strsplit(bbox, ",")[[1]]) != 4){
      stop("Input for bbox must be a string with four comma-seperated coordinate values; e.g. \"52.357, 52.677, 13.228, 13.548\"")
    }
  }

  if(!is.null(wheelchair)&!is.character(wheelchair)){
    stop("Input for wheelchair must be one of the following strings \"yes\", \"limited\", \"no\", \"unknown\".")
  }

  if(!is.null(wheelchair)){
    if(is.character(wheelchair)&!(wheelchair %in% c("yes", "limited", "no", "unknown"))){
      stop("Input for wheelchair must be one of the following strings \"yes\", \"limited\", \"no\", \"unknown\".")
    }
  }
  if (!is.null(page) & !is.numeric(page)) {
    stop("Input for page must be a numeric.")
  }
  if (!is.null(per_page) & !is.numeric(per_page)) {
    stop("Input for per_page must be a numeric.")
  }

  # Chek for internet
  check_internet()
  # Retrieve the results
  res <- GET(base_url, query = compact(args))
  # Check the result
  check_status(res)
  # Get the content and return it as a data.frame
  content_parsed <- parse_wheelmap_content(res)
  # Get meta data
  meta <- extract_wheelmap_metadata(res, content_parsed)
  # extract results
  locations <- extract_wheelmap_locations(meta, content_parsed)

  # format output
  output <- list(meta,locations)
  names(output) <- c("meta_data", "locations")
  # return results in list
  return(output)
}

